From 0286307b13b29902a2987f6b979c540eeded957e Mon Sep 17 00:00:00 2001
From: Felix Hammarstrand <felix.hammarstrand@flir.se>
Date: Fri, 6 Sep 2019 09:56:08 +0200
Subject: [PATCH] add bootlogo functionality

    reads bootlogo from  /boot/bootlogo.bmp.gz and shows on display
    uses mmc active rootfs partition (mmc partition 2 or 3)
---
 board/flir/ec201/ec201.c | 35 +++++++++++++++++++++++++++++++++++
 include/configs/ec201.h  | 12 ++++++++++--
 2 files changed, 45 insertions(+), 2 deletions(-)

diff --git a/board/flir/ec201/ec201.c b/board/flir/ec201/ec201.c
index 776d3dc..85c7fd8 100644
--- a/board/flir/ec201/ec201.c
+++ b/board/flir/ec201/ec201.c
@@ -18,6 +18,7 @@
 #include <mipi_dsi_northwest.h>
 #include <imx_mipi_dsi_bridge.h>
 #include <mipi_dsi_panel.h>
+#include <splash.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -172,4 +173,38 @@ struct display_info_t const displays[] = {{
 		.vmode          = FB_VMODE_NONINTERLACED
 } } };
 size_t display_count = ARRAY_SIZE(displays);
+
+static struct splash_location splash_locations[] = {
+	{
+		.name = "mmc_fs",
+		.storage = SPLASH_STORAGE_MMC,
+		.flags = SPLASH_STORAGE_FS,
+		.devpart = "0:2",
+	}
+};
+
+int splash_screen_prepare(void)
+{
+	//choose which partition to load bootlogo from
+	char * system_active = env_get("system_active");
+	if(system_active)
+	{
+		switch(system_active[6]) //system_active=system1 or system2
+		{
+			case '1':
+				splash_locations[0].devpart[2]='2'; //use mmc partition 2
+				break;
+			case '2':
+				splash_locations[0].devpart[2]='3'; //use mmc partition 3
+				break;
+			default: 
+				printf("splash_screen_prepare: invalid system_active environment: %s \n",system_active); 
+				break;
+		}
+	}
+	
+	return splash_source_load(splash_locations,
+				  ARRAY_SIZE(splash_locations));
+}
+
 #endif
diff --git a/include/configs/ec201.h b/include/configs/ec201.h
index aa20f59..8aa8d03 100644
--- a/include/configs/ec201.h
+++ b/include/configs/ec201.h
@@ -115,8 +115,11 @@
 #define CONFIG_VIDEO_BMP_RLE8
 #define CONFIG_VIDEO_BMP_LOGO
 #define CONFIG_IMX_VIDEO_SKIP
-
 #define CONFIG_HX8394
+
+#define CONFIG_SPLASH_SOURCE
+#define CONFIG_VIDEO_BMP_GZIP
+#define CONFIG_SYS_VIDEO_LOGO_MAX_SIZE 1024*1024
 #endif
 
 /* Override imx standard behavior when loading u-boot from usb */
@@ -381,7 +384,11 @@
 				"bootaux ${m4_addr};" \
 			"fi;\0"
 
-
+#define CONFIG_SPLASH_IMAGE_ENV \
+	"splashimage=0x67000000\0" \
+	"splashsource=mmc_fs\0" \
+	"splashfile=/boot/bootlogo.bmp.gz\0" \
+	"panel=TRULY-VGA-SHERLOCK\0"
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	CONFIG_DEFAULT_NETWORK_SETTINGS \
@@ -401,6 +408,7 @@
 	CONFIG_SELECT_BOOT_PARTITION_ENV \
 	CONFIG_MMC_BOOT_FLIR_ENV\
 	CONFIG_MMC_BOOT_LEGACY_ENV \
+	CONFIG_SPLASH_IMAGE_ENV \
 	""	/* end line */
 
 
